diff --git a/src/parse_rawml.c b/src/parse_rawml.c
index 6b6c9d7..41ba9a3 100644
--- a/src/parse_rawml.c
+++ b/src/parse_rawml.c
@@ -107,7 +107,7 @@ MOBI_RET mobi_search_links_kf7(MOBIResult *result, const unsigned char *data_sta
                     result->value[i++] = (char) *data++;
                 }
                 /* self closing tag '/>' */
-                if (*(data - 1) == '/' && *data == '>') {
+                if (data <= data_end && *(data - 1) == '/' && *data == '>') {
                     --data; --i;
                 }
                 result->end = data;
