diff --git a/mrbgems/mruby-compiler/core/codegen.c b/mrbgems/mruby-compiler/core/codegen.c
index 5ff33b870..3e9a153c2 100644
--- a/mrbgems/mruby-compiler/core/codegen.c
+++ b/mrbgems/mruby-compiler/core/codegen.c
@@ -1136,6 +1136,7 @@ gen_int(codegen_scope *s, uint16_t dst, mrb_int i)
 static mrb_bool
 gen_uniop(codegen_scope *s, mrb_sym sym, uint16_t dst)
 {
+  if (no_peephole(s)) return FALSE;
   struct mrb_insn_data data = mrb_last_insn(s);
   mrb_int n;
 